File: jvm/core/src/main/scala/ml/dmlc/tvm/types/TVMType.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ml.dmlc.tvm.types

object TVMType {
  implicit def str2Type(dtype: String): TVMType = TVMType(dtype)

  val CODE2STR = Map(
    0 -> "int",
    1 -> "uint",
    2 -> "float",
    4 -> "handle"
  )

  val INT = 0
  val UINT = 1
  val FLOAT = 2
  val HANDLE = 4

  def apply(typeStr: String, lanes: Int = 1): TVMType = {
    new TVMType(typeStr, lanes)
  }
}

class TVMType(val typeStr: String, val lanes: Int = 1) {
  private val (typeCodeTemp, bitsTemp) =
    if (typeStr.startsWith("int")) {
      (0, typeStr.substring(3).toInt)
    } else if (typeStr.startsWith("uint")) {
      (1, typeStr.substring(4).toInt)
    } else if (typeStr.startsWith("float")) {
      (2, typeStr.substring(5).toInt)
    } else if (typeStr.startsWith("handle")) {
      (4, 64)
    } else {
      throw new IllegalArgumentException("Do not know how to handle type " + typeStr)
    }

  val typeCode = typeCodeTemp
  val bits = if (bitsTemp == 0) 32 else bitsTemp
  if ((bits & (bits - 1)) != 0 || bits < 8) {
    throw new IllegalArgumentException("Do not know how to handle type " + typeStr)
  }

  def numOfBytes: Int = {
    bits / 8
  }

  override def hashCode: Int = {
    (typeCode << 16) | (bits  << 8) | lanes
  }

  override def equals(other: Any): Boolean = {
    if (other != null && other.isInstanceOf[TVMType]) {
      val otherInst = other.asInstanceOf[TVMType]
      (bits == otherInst.bits) && (typeCode == otherInst.typeCode) && (lanes == otherInst.lanes)
    } else {
      false
    }
  }

  override def toString: String = {
    val str = s"${TVMType.CODE2STR(typeCode)}$bits"
    if (lanes != 1) {
      str + lanes
    } else {
      str
    }
  }
}

File: jvm/core/src/main/java/org/apache/tvm/TVMType.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.tvm;

public class TVMType {
  public static final int INT = 0;
  public static final int UINT = 1;
  public static final int FLOAT = 2;
  public static final int HANDLE = 4;

  public final int typeCode;
  public final int bits;
  public final int numOfBytes;
  public final int lanes;

  /**
   * TVMType constructor.
   * @param typeStr type name, e.g., "float32", "float64", "uint8", etc.
   * @param lanes Tensor lanes.
   */
  public TVMType(String typeStr, int lanes) {
    this.lanes = lanes;
    int bitsTemp = 0;
    if (typeStr.startsWith("int")) {
      typeCode = INT;
      bitsTemp = Integer.parseInt(typeStr.substring(3));
    } else if (typeStr.startsWith("uint")) {
      typeCode = UINT;
      bitsTemp = Integer.parseInt(typeStr.substring(4));
    } else if (typeStr.startsWith("float")) {
      typeCode = FLOAT;
      bitsTemp = Integer.parseInt(typeStr.substring(5));
    } else if (typeStr.startsWith("handle")) {
      typeCode = HANDLE;
      bitsTemp = 64;
    } else {
      throw new IllegalArgumentException("Do not know how to handle type " + typeStr);
    }
    bits = (bitsTemp == 0) ? 32 : bitsTemp;
    if ((bits & (bits - 1)) != 0 || bits < 8) {
      throw new IllegalArgumentException("Do not know how to handle type " + typeStr);
    }
    numOfBytes = bits / 8;
  }

  public TVMType(String typeStr) {
    this(typeStr, 1);
  }

  @Override public int hashCode() {
    return (typeCode << 16) | (bits  << 8) | lanes;
  }

  @Override public boolean equals(Object other) {
    if (other != null && other instanceof TVMType) {
      TVMType otherInst = (TVMType) other;
      return (bits == otherInst.bits)
        && (typeCode == otherInst.typeCode) && (lanes == otherInst.lanes);
    }
    return false;
  }

  @Override public String toString() {
    String typeCodeStr;
    switch (typeCode) {
      case INT:
        typeCodeStr = "int";
        break;
      case UINT:
        typeCodeStr = "uint";
        break;
      case FLOAT:
        typeCodeStr = "float";
        break;
      case HANDLE:
        typeCodeStr = "handle";
        break;
      default:
        typeCodeStr = "Unknown";
        break;
    }
    String str = typeCodeStr + bits;
    if (lanes != 1) {
      str += lanes;
    }
    return str;
  }
}
